R4R ADR 007 / 008: Specialization Groups Abstraction / CERT Group (#4827)
cosmos/cosmos-sdk: docs/architecture/adr-007-specialization-groups.md

# ADR 007: Specialization Groups

## Changelog

- 2019 Jul 31 : Initial Conception

## Context

This idea was first conceived of in order to fulfill the use case of the

> creation of a decentralized Computer Emergency Response Team (CERT), Whose

Reviewer: 😃 
```suggestion
creation of a decentralized Computer Emergency Response Team (dCERT), Whose
```

Author: LOVE IT

Author: 🍰

members would be elected by the Cosmos Community and would fulfill the role of
coordinating the Cosmos community under an emergency situations. This thinking
can be further abstracted into the conception of "blockchain specialization
groups". 

The creation of these groups are the beginning of specialization capabilities
within a wider blockchain community which could be used to enable a certain
level of delegated responsibilities. Examples of specialization which could be

> beneficial to a blockchain community include: code auditing, emergency response,

Reviewer 2: Do you have an example of emergency response? Also an example to show where a CERT team could handle an emergency more effectively than direct governance vote would be useful in providing context.

Author: I've added more discussion (but not here) to the Context section of ADR 008 (in this PR) which should fully describe what you're looking for

code development etc. This type of community organization paves the way for
individual stakeholders to delegate votes by issue type, if in the future
governance proposals include a field for issue type. 

## Decision

A specialization group can be broadly broken down into
the following functions: 
 - Membership Admittance
   - via governance

>    - via special appointment

Reviewer: can you clarify this item?

Author: These are examples of potentially implementable situations, I'll clarify

Author: Added a discussion paragraph talking about this stuff and removed these bullets from under `Membership Admittance`

 - Membership Acceptance
 - Membership Revocation
   - (probably) Without Penalty 
     - member steps down (self-Revocation)
     - replaced by new member from governance
   - (probably) With Penalty 
     - due to breach of soft-agreement (determined through governance)
     - due to breach of hard-agreement (determined by code) 
 - Execution of Duties

>    - Special transactions which only execute for members of a specialization group

Reviewer: for example?

Author: added an example

 - Compensation
   - Group compensation (further distribution decided by the specialization group) 
   - Individual compensation for all constituents of a group from the
     greater community

Election of the members of a specialization group can happen in a wide variety
of ways and be subject to an arbitrary number of associated rules. For its
initial implementation as a part of this ADR we recommend that a general

> election abstraction (`Electionator`) is provided as well as a basic

Reviewer: lol this sounds like Terminator but for governance

implementation of that abstraction which allows for a continuous election of
members of a specialization group. 

``` golang
// The Electionator abstraction covers the concept space for 
// a wide variety of election kinds.  
type Electionator interface {
    
    // is the election object accepting votes.
    Active() bool 

    // functionality to execute for when a vote is cast in this election, here
    // the vote field is anticipated to be marshalled into a vote type used 
    // by an election. 

>     Vote(addr sdk.AccAddress, vote []byte) 

Reviewer: and the proposal id?

Author: I don't think that the proposal id is required here, because this is a vote _on_ this Electionator, each electionator is not anticipated to have multiple proposals within it.

Reviewer: 👍 that needs to be clarified in the comments too

Author: I added more discussion to this effect here

    // here lies all functionality to authenticate and execute changes for
    // when a member accepts being elected

>     Accept(sdk.AccAddress) 

Reviewer: `Accept` is too broad. Maybe consider something like 
```suggestion
    AcceptElection(sdk.AccAddress) 
```

Author: added

    // Register a revoker object
    RegisterRevoker(Revoker)

    // No more revokers may be registered after this function is called
    SealRevokers()

    // register hooks to call when an election actions occur
    RegisterHooks(ElectionatorHooks) 

    // query for the current winner(s) of this election based on arbitrary
    // election ruleset
    QueryWinners() []sdk.AccAddress 

    // query metadata for an address in the election this 
    // could include for example position that an address
    // is being elected for within a group   
    // 
    // this metadata may be directly related to 
    // voting information and/or privileges enabled
    // to members within a group. 
    QueryMetadata(sdk.AccAddress) []byte
}

> type ElectionatorHooks interface {

Reviewer: important to note that these hooks trigger events

Author: Added a comment

>     VoteCast(addr sdk.AccAddress, vote []byte)

Reviewer: these 3 hooks are missing the proposal ID. Otherwise, you won't be able to distinguish similar events from one proposal to another

Author: So there isn't the concept of proposalIDs here - these vote belong _exclusively_ to the Electionator (which itself can be thought of as a single proposal)

Reviewer: I don't follow. Can there be multiple proposals to register more members?

Author: There are no proposals here. Just votes which pertain specifically to one electionator. Anyone can create and send a vote to the electionator item which will presumably attempt to marshal those bytes into a particular struct and apply the vote information in some arbitrary way. There can be multiple Electionators within the Cosmos-Hub for multiple specialization groups, votes would need to be routed to the Electionator _upstream_ of here.

// and evidence based revoking, etc. Revokers types may be created and
// reused for different election types. 
// 
// When revoking the "cause" bytes may be arbitrarily marshalled into evidence,
// memos, etc.
type Revoker interface {
    RevokeName() string      // identifier for this revoker type 
    RevokeMember(addr sdk.AccAddress, cause []byte) (successful bool)
}
```

Certain level of commonality likely exists between the existing code within
`x/governance` and required functionality of elections. This common
functionality should be abstracted during implementation. Similarly for each 
vote implementation client CLI/REST functionality should be abstracted
to be reused for multiple elections. 

The specialization group abstraction firstly extends the `Electionator`
but also further defines traits of the group. 

``` golang
type SpecializationGroup interface {
    Electionator 
    GetName() string
    GetDescription() string 

    // general soft contract the group is expected
    // to fulfill with the greater community
    GetContract() string

    // messages which can be executed by the members of the group

> 	Handler(ctx sdk.Context, msg sdk.Msg) sdk.Result

Reviewer: indentation is off

    // logic to be executed at endblock, this may for instance
    // include payment of a stipend to the group members
    // for participation in the security group.   

> 	EndBlocker(ctx sdk.Context)

Reviewer: ditto

}
```

## Status

> Proposed

> ## Consequences

Reviewer: it'd be helpful to mention if this requires to do further changes to the current governance module (if any)

Author: added ` - improve abstractions in `x/gov/` such that they can be used with specialization groups` to Positives though

# ADR 007: Specialization Groups

## Changelog

> - 2019 Jul 31 : Initial Conception

Reviewer 3: ```suggestion
- 2019 Jul 31: Initial Draft
```

## Context

This idea was first conceived of in order to fulfill the use case of the
creation of a decentralized Computer Emergency Response Team (dCERT), Whose
members would be elected by the Cosmos Community and would fulfill the role of

> coordinating the Cosmos community under an emergency situations. This thinking

Reviewer 3: ```suggestion
coordinating the Cosmos community under emergency situations. This thinking
```

can be further abstracted into the conception of "blockchain specialization
groups". 

The creation of these groups are the beginning of specialization capabilities
within a wider blockchain community which could be used to enable a certain
level of delegated responsibilities. Examples of specialization which could be
beneficial to a blockchain community include: code auditing, emergency response,
code development etc. This type of community organization paves the way for
individual stakeholders to delegate votes by issue type, if in the future
governance proposals include a field for issue type. 

## Decision

A specialization group can be broadly broken down into the following functions

> (herein containing examples): 

Reviewer 3: ```suggestion
(herein containing examples):

```

 - Membership Admittance
 - Membership Acceptance
 - Membership Revocation
   - (probably) Without Penalty 
     - member steps down (self-Revocation)
     - replaced by new member from governance

>    - (probably) With Penalty 

Reviewer 3: How do you propose penalties be executed? What is the general idea and mechanism behind this? Should we allude to this at all in the ADR?

Author: These are just general ideas as to mechanisms which many specialization groups will probably want - this is mostly a high level overview. I'm doubtful that there is a way to easily abstract this functionality into the interface level, however if that's the case we can add that as an addendum to this ADR. Otherwise this functionality is intended to be built by the instance of the interfaces proposed.  (so I don't think we should allude to any more specifics than we've given, at least for now)

     - due to breach of soft-agreement (determined through governance)
     - due to breach of hard-agreement (determined by code) 
 - Execution of Duties
   - Special transactions which only execute for members of a specialization
     group (for example, dCERT members voting to turn off transaction routes in
     an emergency scenario) 
 - Compensation
   - Group compensation (further distribution decided by the specialization group) 
   - Individual compensation for all constituents of a group from the
     greater community

Membership admittance to a specialization group could take place over a wide
variety of mechanisms. The most obvious example is through a general vote among
the entire community, however in certain systems a community may want to allow
the members already in a specialization group to internally elect new members,
or maybe the community may assign a permission to a particular specialization
group to appoint members to other 3rd party groups. The sky is really the limit

> as to how membership admittance can be structured. We attempt to capture these

Reviewer 3: ```suggestion
as to how membership admittance can be structured. We attempt to capture
```

    // when a member accepts being elected
    AcceptElection(sdk.AccAddress) 

    // Register a revoker object
    RegisterRevoker(Revoker)

    // No more revokers may be registered after this function is called
    SealRevokers()

    // register hooks to call when an election actions occur
    RegisterHooks(ElectionatorHooks) 

    // query for the current winner(s) of this election based on arbitrary
    // election ruleset
    QueryWinners() []sdk.AccAddress 

    // query metadata for an address in the election this 
    // could include for example position that an address
    // is being elected for within a group   
    // 
    // this metadata may be directly related to 
    // voting information and/or privileges enabled
    // to members within a group. 
    QueryMetadata(sdk.AccAddress) []byte
}

// ElectionatorHooks, once registered with an Electionator, 
// trigger execution of relevant interface functions when 
// Electionator events occur. 
type ElectionatorHooks interface {
    AfterVoteCast(addr sdk.AccAddress, vote []byte)
    AfterMemberAccepted(addr sdk.AccAddress)
    AfterMemberRevoked(addr sdk.AccAddress, cause []byte)
}

> // Revoker defines the function required for an membership revocation rule-set

Reviewer 3: ```suggestion
// Revoker defines the function required for a membership revocation rule-set
```

// used by a specialization group. This could be used to create self revoking,
// and evidence based revoking, etc. Revokers types may be created and
// reused for different election types. 
// 
// When revoking the "cause" bytes may be arbitrarily marshalled into evidence,
// memos, etc.

> type Revoker interface {

Reviewer 2: There should be a way for governance to disband an entire specialization group if they are caught colluding maliciously

Author: added:  `emergency disband of the entire dCERT group (ex. for colluding maliciously)` to ADR 008 - I think this is not a generalized specialization group problem but relevant for dCERT

Author: Also added this 
```
Additionally in the case of an emergency situation of a colluding and malicious
dCERT group, the community needs the capability to disband the entire dCERT
group and likely fully slash them. This could be achieved though a special new
proposal type (general governance) which would halt the functionality of the
dCERT group until the proposal was concluded. This special proposal type would
likely need to also have a fairly large wager which could be slashed if the
proposal creator was malicious. 
```

# ADR 007: Specialization Groups

## Changelog

- 2019 Jul 31: Initial Draft

## Context

This idea was first conceived of in order to fulfill the use case of the

> creation of a decentralized Computer Emergency Response Team (dCERT), Whose

Reviewer 3: ```suggestion
creation of a decentralized Computer Emergency Response Team (dCERT), whose
```

> members would be elected by the Cosmos Community and would fulfill the role of

Reviewer 4: What is the "Cosmos Community"? Are you referring specifically to the Atom holder set?

> members would be elected by the Cosmos Community and would fulfill the role of

Reviewer 3: Maybe this instead?

```suggestion
members would be elected by a governing community and would fulfill the role of
```

> coordinating the Cosmos community under emergency situations. This thinking

Reviewer 3: ```suggestion
coordinating the community under emergency situations. This thinking
```

    // the vote field is anticipated to be marshalled into a vote type used 
    // by an election. 
    // 
    // NOTE There are no explicit ids here. Just votes which pertain specifically
    // to one electionator. Anyone can create and send a vote to the electionator item
    // which will presumably attempt to marshal those bytes into a particular struct
    // and apply the vote information in some arbitrary way. There can be multiple
    // Electionators within the Cosmos-Hub for multiple specialization groups, votes
    // would need to be routed to the Electionator upstream of here.
    Vote(addr sdk.AccAddress, vote []byte) 

    // here lies all functionality to authenticate and execute changes for
    // when a member accepts being elected
    AcceptElection(sdk.AccAddress) 

    // Register a revoker object
    RegisterRevoker(Revoker)

    // No more revokers may be registered after this function is called
    SealRevokers()

    // register hooks to call when an election actions occur
    RegisterHooks(ElectionatorHooks) 

    // query for the current winner(s) of this election based on arbitrary
    // election ruleset

>     QueryWinners() []sdk.AccAddress 

Reviewer 3: I would recommend changing the nomenclature here. Perhaps `QueryElected()`

    // query metadata for an address in the election this 
    // could include for example position that an address
    // is being elected for within a group   
    // 
    // this metadata may be directly related to 
    // voting information and/or privileges enabled
    // to members within a group. 
    QueryMetadata(sdk.AccAddress) []byte
}

// ElectionatorHooks, once registered with an Electionator, 
// trigger execution of relevant interface functions when 
// Electionator events occur. 
type ElectionatorHooks interface {
    AfterVoteCast(addr sdk.AccAddress, vote []byte)
    AfterMemberAccepted(addr sdk.AccAddress)
    AfterMemberRevoked(addr sdk.AccAddress, cause []byte)
}

// Revoker defines the function required for a membership revocation rule-set
// used by a specialization group. This could be used to create self revoking,
// and evidence based revoking, etc. Revokers types may be created and
// reused for different election types. 
// 
// When revoking the "cause" bytes may be arbitrarily marshalled into evidence,
// memos, etc.
type Revoker interface {
    RevokeName() string      // identifier for this revoker type 

>     RevokeMember(addr sdk.AccAddress, cause []byte) (successful bool)

Reviewer 3: This should rather return an error.

functionality should be abstracted during implementation. Similarly for each 
vote implementation client CLI/REST functionality should be abstracted
to be reused for multiple elections. 

The specialization group abstraction firstly extends the `Electionator`
but also further defines traits of the group. 

``` golang
type SpecializationGroup interface {
    Electionator 
    GetName() string
    GetDescription() string 

    // general soft contract the group is expected
    // to fulfill with the greater community
    GetContract() string

    // messages which can be executed by the members of the group
    Handler(ctx sdk.Context, msg sdk.Msg) sdk.Result

    // logic to be executed at endblock, this may for instance
    // include payment of a stipend to the group members
    // for participation in the security group.   
    EndBlocker(ctx sdk.Context)
}
```

## Status

> Proposed

## Consequences

### Positive

 - increases specialization capabilities of a blockchain
 - improve abstractions in `x/gov/` such that they can be used with specialization groups

### Negative

 - could be used to increase centralization within a community

### Neutral

## References

>   (dCERT ADR)[./adr-008-dCERT-group.md]

Reviewer 3: ```suggestion
* (dCERT ADR)[./adr-008-dCERT-group.md]
```
